geode-dunit/src/main/java/org/apache/geode/management/internal/cli/commands/NestedQueryClassCastExceptionFailureDUnitTest.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to You under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package org.apache.geode.management.internal.cli.commands;
+
+import static org.apache.geode.distributed.ConfigurationProperties.SERIALIZABLE_OBJECT_FILTER;
+import static org.junit.Assert.assertEquals;
+
+import java.util.Properties;
+import java.util.TreeMap;
+
+import org.junit.Rule;
+import org.junit.Test;
+
+import org.apache.geode.cache.Cache;
+import org.apache.geode.cache.CacheFactory;
+import org.apache.geode.cache.Region;
+import org.apache.geode.cache.RegionShortcut;
+import org.apache.geode.cache.query.QueryService;
+import org.apache.geode.cache.query.SelectResults;
+import org.apache.geode.cache.server.CacheServer;
+import org.apache.geode.pdx.ReflectionBasedAutoSerializer;
+import org.apache.geode.test.dunit.Host;
+import org.apache.geode.test.dunit.VM;
+import org.apache.geode.test.dunit.rules.ClusterStartupRule;
+import org.apache.geode.test.version.VersionManager;
+
+public class NestedQueryClassCastExceptionFailureDUnitTest {
+  @Rule
+  public ClusterStartupRule cluster = new ClusterStartupRule();
+
+  @Test
+  public void classCastExceptionShouldNotBeThrownWhileExecutionNestedQueries() {
+    Host host = Host.getHost(0);
+    VM server = host.getVM(VersionManager.CURRENT_VERSION, 0);
+    server.invoke(() -> {
+      Properties properties = new Properties();
+      properties.put(SERIALIZABLE_OBJECT_FILTER,
+          "org.apache.geode.management.internal.cli.commands.*");
+      Cache cache = new CacheFactory(properties)
+          .setPdxSerializer(new ReflectionBasedAutoSerializer(
+              "org.apache.geode.management.internal.cli.commands.*"))
+          .create();
+
+      CacheServer cacheServer = cache.addCacheServer();
+      cacheServer.setPort(0);
+      cacheServer.start();
+      Region region = cache.createRegionFactory(RegionShortcut.REPLICATE).create("product");
```

Comment:
Nitpick: this chunk of code can be replaced with something "easier" to read:

```
  @Rule
  public ServerStarterRule server =
      new ServerStarterRule()
          .withRegion(RegionShortcut.REPLICATE, "product")
          .withProperty(SERIALIZABLE_OBJECT_FILTER, "org.apache.geode.management.internal.cli.commands.*")
          .withPdxSerializer(new ReflectionBasedAutoSerializer("org.apache.geode.management.internal.cli.commands.*"))
          .withAutoStart();

  @Test
  public void classCastExceptionShouldNotBeThrownWhileExecutionNestedQueries() throws Exception {
      Cache cache = server.getCache();
      Region<Integer, Product> region = cache.getRegion("product");
      (...)
}
```

Reply:
I agree, but I personally prefer using creating the setup as we say in our documentation. Previously I had issues with Rules. I will try to refactor the tests better from now on to make it more readable